packages/ni.measurementlink.discovery.v1.client/src/ni/measurementlink/discovery/v1/client/_client.py

```diff
@@ -0,0 +1,269 @@
+"""Client for accessing the NI Discovery Service."""
+
+from __future__ import annotations
+
+import logging
+import threading
+from collections.abc import Sequence
+
+import grpc
+from ni.measurementlink.discovery.v1 import discovery_service_pb2, discovery_service_pb2_grpc
+from ni_grpc_extensions.channelpool import GrpcChannelPool  # type: ignore[import-untyped]
+
+from ni.measurementlink.discovery.v1.client._info import ServiceInfo
+from ni.measurementlink.discovery.v1.client._support import (
+    _get_discovery_service_address,
+)
+from ni.measurementlink.discovery.v1.client._types import (
+    ComputeNodeDescriptor,
+    ServiceLocation,
+)
+
+SERVICE_PROGRAMMINGLANGUAGE_KEY = "ni/service.programminglanguage"
+_logger = logging.getLogger(__name__)
+
+
+class DiscoveryClient:
+    """Client for accessing the NI Discovery Service."""
+
+    def __init__(
+        self,
+        stub: discovery_service_pb2_grpc.DiscoveryServiceStub | None = None,
+        *,
+        grpc_channel_pool: GrpcChannelPool | None = None,
+    ) -> None:
+        """Initialize the discovery client.
+
+        Args:
+            stub: An optional discovery service gRPC stub for unit testing.
+
+            grpc_channel_pool: An optional gRPC channel pool (recommended).
+        """
+        self._initialization_lock = threading.Lock()
+        self._grpc_channel_pool = grpc_channel_pool
+        self._stub = stub
+        self._registration_id = ""
```

Comment:
Add `__slots__` to prevent shenanigans. `__weakref__` is fine but not `__dict__`.

Reply:
Include `__weakref__` in `__slots__`